File: add_prog_bin_path_to_ENV_PATH_.py
#coding:gbk
#2011_02_20 add by greshem
#把常见的目录都添加进去了.
#添加java 的一些东西 classpath JAVA_HOME 等. 
import sys
import site
import os
import _winreg

HKCU = _winreg.HKEY_CURRENT_USER
ENV = "Environment"
PATH = "PATH"
DEFAULT = u"%PATH%"

def add_path_to_ENV_PATH(add_path):
	
    with _winreg.CreateKey(HKCU, ENV) as key:
        try:
            envpath = _winreg.QueryValueEx(key, PATH)[0]
        except WindowsError:
            envpath = DEFAULT

        #默认添加的方式.  
        paths = [envpath]
        if add_path and add_path not in envpath and os.path.isdir(add_path):
            paths.append(add_path)

        envpath = os.pathsep.join(paths)
        _winreg.SetValueEx(key, PATH, 0, _winreg.REG_EXPAND_SZ, envpath)
        return paths, envpath

def main():
    paths, envpath = add_path_to_ENV_PATH("c:\\bin");
    paths, envpath = add_path_to_ENV_PATH("c:\\cygwin\\bin");
    paths, envpath = add_path_to_ENV_PATH("C:\\Ruby192\\bin");

    paths, envpath = add_path_to_ENV_PATH("c:\\Perl\\bin");
    paths, envpath = add_path_to_ENV_PATH("C:\\q******************************n_bin");
    paths, envpath = add_path_to_ENV_PATH("C:\\Program Files\\Vim\\vim72");

	#C:\Program Files\Bakefile
    paths, envpath = add_path_to_ENV_PATH("C:\\Program Files\\Bakefile");

	#C:\Program Files\Wireshark
    paths, envpath = add_path_to_ENV_PATH("C:\\Program Files\\Wireshark");

	#C:\\Program Files\\Ext2Fsd
    paths, envpath = add_path_to_ENV_PATH("C:\\Program Files\\Ext2Fsd");


    #C:\Program Files\Subversion\bin
    paths, envpath = add_path_to_ENV_PATH("C:\\Program Files\\Subversion\\bin");
    print _winreg.ExpandEnvironmentStrings(envpath)

if __name__ == '__main__':
    main()
    print "input any keys to quit  \n";
    if sys.platform =='win32':
        print sys.stdin.readline()[:-1];

File: add_prog_bin_path_to_ENV_PATH.pl
#!/usr/bin/perl
#2011_08_10   星期三   add by greshem
#还是要用 QueryValueEx 否则还是不行. 
#之后所有的环境变量的维护都可以在这个程序里面了了
sub add_path_to_ENV($$)
{
	use Win32::Registry;
	(my $sub_key,my  $in_value)=@_;
	$in_value_pattern=$in_value;
	$in_value_pattern=~s/\\/\\\\/g;

	my $Register = "Environment"; #没有子key所以为空
	my $tmp, $tmp2; 
	my $value;
	$HKEY_CURRENT_USER->Open($Register , $hkey) || die $!;
	$hkey->QueryValueEx($sub_key,undef, $value ) or die $!;
	print $value."\n";
	my @array=split(/;/, $value);
	my @tmp = grep{/$in_value_pattern/} @array;
	if(scalar(@tmp) >=1)
	{
		print "$in_value 已经添加了, 不必再添加\n";
	}
	else
	{
		push (@array, $in_value);
	}
	$last_value=join(";", @array);
	$hkey->SetValueEx($sub_key,undef,REG_SZ, $last_value);
	$hkey->Close();
	print "最后的字符串是: $last_value \n";

}

my $input_dir=shift or warn("usage: $0 input_dir \n");

if( -d $input_dir)
{
	add_path_to_ENV("PATH","$input_dir");
	print (" rundll32.exe shell32.dll,Control_RunDLL sysdm.cpl,,3 \n");
	system (" rundll32.exe shell32.dll,Control_RunDLL sysdm.cpl,,3 \n");
}
else
{
	init_system();
}

#添加常用的地址. 
sub init_system()
{
	add_path_to_ENV("PATH","c:\\bin");
	add_path_to_ENV("PATH","C:\\Python26");
	add_path_to_ENV("PATH","c:\\cygwin\\bin");
	add_path_to_ENV("PATH","C:\\Ruby192\\bin");
	add_path_to_ENV("PATH","c:\\Perl\\bin");
	add_path_to_ENV("PATH","C:\\Program Files\\Vim\\vim72");
	add_path_to_ENV("PATH","C:\\Program Files\\Bakefile");
	add_path_to_ENV("PATH","C:\\Program Files\\Wireshark");
	add_path_to_ENV("PATH","C:\\Program Files\\Ext2Fsd");
	add_path_to_ENV("PATH","C:\\Program Files\\Subversion\\bin");

	add_path_to_ENV("PATH", "C:\\MPlayer_Windows");
	add_path_to_ENV("PATH", "C:\\Program Files\\HTML Help Workshop");
	add_path_to_ENV("PATH", "C:\\Program Files\\WinCDEmu");
	add_path_to_ENV("PATH", "C:\\Program Files\\WinRAR");
	add_path_to_ENV("PATH", "C:\\Program Files\\Debugging Tools for Windows (x86)\\");
	add_path_to_ENV("PATH", "C:\\Leakdiag\\");

	add_path_to_ENV("PATH", "C:\\Program Files\\Microsoft Visual Studio\\COMMON\\MSDev98\\Bin");
}
